pkg/kubelet/pod_workers.go

```diff
@@ -244,13 +244,33 @@ type podSyncStatus struct {
 	ctx context.Context
 	// cancelFn if set is expected to cancel the current sync*Pod operation.
 	cancelFn context.CancelFunc
-	// working is true if a pod worker is currently in a sync method.
-	working bool
+
 	// fullname of the pod
 	fullname string
 
+	// working is true if an update is pending or being worked by a pod worker
+	// goroutine.
+	working bool
+	// pendingUpdate is the updated state the pod worker should observe. It is
+	// cleared when a pod worker reads it. A new update may always replace
+	// a pending update as the pod worker does not guarantee that all intermediate
+	// states are synced to a worker, only the most recent. This state should not
+	// be visible to downstream components until a pod worker has begun processing
+	// it.
+	pendingUpdate *UpdatePodOptions
+	// lastUpdate is the last version of the pod's state known to the worker.
```

Comment:
It's ok to race with syncPods because it's long running - a syncPod call could take minutes or hours (because all hooks are synchronous today).  So it's fine for downstream components to get ahead of the pod during setup (this pod will definitely start as soon as the goroutine gets moving), as long as they don't do things for pods that won't start ever, and as long as they don't start tearing down until syncTerminatingPods exits (which we have logic for).

There is still racing, but it should be the controlled kind of race (everyone can set this pod up in parallel).  We should probably document that in the getters once they are added.

Right now this is probably "activeSyncUpdate" - not sure how clean that name is.

Reply:
That makes sense and most likely if those components are relying on pod manager today they are racy as well because pod manager was not previously synchronized with pod worker. I agree it would be useful to document this distinction if it is important for some downstream component.

Regarding naming, `activeSyncUpdate` makes sense to me, but we can also keep existing name and just document that the lastUpdate may still be in progress.

Reply:
Or even `activeUpdate`?  `syncUpdate`?

Reply:
+1 on `activeUpdate`, keep it simple :)